pkg/manager/manager_test.go

```diff
@@ -1044,6 +1045,44 @@ var _ = Describe("manger.Manager", func() {
 				}))).NotTo(Succeed())
 			})
 
+			It("should not return runnables context.Canceled errors", func() {
+				Expect(options.Logger).To(BeZero(), "this test overrides Logger")
+
+				logs := []string{}
+				options.Logger = funcr.NewJSON(func(object string) {
+					logs = append(logs, object)
```

Comment:
This is a datarace which is why the tests are failing. You need to protect `logs` with a mutex or some such

Reply:
A mutex here solved a race triggered by `append`, but that revealed a race between this line and the assertion on L1091 during the LeaderElection suite. My guess is that this goroutine is to blame: https://github.com/kubernetes-sigs/controller-runtime/blob/21368602d84b669eacfb36ec49b10a665b2d4d49/pkg/manager/internal.go#L423-L427


😞 I added a `time.Sleep` for that suite for now.

Reply:
So the idea is that currently with leaderelection engageStopProcedure is still writing the the logger while the test is asserting on the logs?

Reply:
My guess would be because the RunnableFunc is not a LeaderElectionRunnable. So it won't wait for leaderelection to close start

Reply:
Okay reproduced locally. ^^ doesn't make sense as everything that doesn't implement LeaderElectionRunnable is treated like it needs leader election

Reply:
I think the problem is that nothing is actually waiting for this go routine to return:
```go
	go func() {
		for {
			// Closing in the for loop is required to avoid race conditions between
			// the closure of all internal procedures and making sure to have a reader off the error channel.
			closeOnce.Do(func() {
				// Cancel the internal stop channel and wait for the procedures to stop and complete.
				close(cm.internalProceduresStop)
				cm.internalCancel()
			})
			select {
			case err := <-cm.errChan:
				if !errors.Is(err, context.Canceled) {
					cm.logger.Error(err, "error received after stop sequence was engaged")
				}
			case <-stopComplete:
				return
			}
		}
	}()
```

While we call `defer close(stopComplete)` at the end of Start, this only means that the go routine will return pretty soon after, but we don't wait for it before we return. So it makes sense that the goroutine still writes to the logger even after Start returned

Reply:
I think the reason why it only occurs with leader election is because in that case we get one more log message:
<img width="783" alt="image" src="https://github.com/kubernetes-sigs/controller-runtime/assets/4662360/4b5cc350-0e49-455f-9900-9d1dab7f6e93">
(the other log messages are all produced "synchronously" within engageStopProcedure)

Reply:
@cbandy @alvaroaleman If that makes sense to you both, let's add a godoc comment before the sleep and I think we're good

Reply:
🤔 The fact that `Start` aggregates its `err` with `stopErr` makes me think it expects `engageStopProcedure` to return the outcome of stopping. Why does it inject a channel at all?

Should `engageStopProcedure` wait for the `errChan`-draining goroutine to finish?

Reply:
> 🤔 The fact that Start aggregates its err with stopErr makes me think it expects engageStopProcedure to return the outcome of stopping. Why does it inject a channel at all?

Not sure. I think the goal of this case `case err := <-cm.errChan:` in `engageStopProcedure` is mostlyl to avoid running into deadlocks because the errChan blocks.

> Should engageStopProcedure wait for the errChan-draining goroutine to finish?

With the way it currently works it can't because this defer (`defer close(stopComplete)`) in Start is called too late.

Reply:
I guess we could wait in the Start func that the `errChan`-draining goroutine also finishes.

It would be a bit nicer, but I'm not sure if I want to introduce even more complexity into all of this. It's already pretty had to reason about it.